### DISCUSSION TITLE
Authentication to remote AAD protected AKS

### DISCUSSION THREAD

**Discussion by radirobi** [fluxcd/flux2#4153] [Q&A]
Hey everyone,

we would like to manage remote Kubernetes clusters from a central Kubernetes cluster where the FluxCD instance is running on.

The remote clusters are **Azure Kubernetes Services** (AKS) clusters and those clusters are **protected by Azure Active Directory.** To be able to authenticate to those clusters you need to use the tool called **kubelogin**. 

I am aware of the property called **kubeConfig** in Kustomization and HelmRelease manifests which should point to a secret which contains the credentials.

What is the ideal way to handle this kind of scenario? One thing that comes to my mind is installing the kubelogin binary into the proper controllers via some initContainers. Any ideas are welcomed.

**Comment by makkes** [comment]
You can create a ServiceAccount and RBAC rules, then [create a token](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-an-api-token-for-a-serviceaccount) that you [use in the kubeconfig](https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/#define-clusters-users-and-contexts).

**Reply by radirobi** [reply]
Hey @makkes,

thanks for taking a look at my problem. Yes, that would be an option, however our AKS cluster has the **[disable Local accounts](https://learn.microsoft.com/en-us/azure/aks/manage-local-accounts-managed-azure-ad#disable-local-accounts)** flag turned on.

It means that the traditional users, service accounts can not be used for authentication. Only those entities are valid which are recognized by the Azure Active Directory. Azure Active Directory does not know anything about pure Kubernetes service accounts. 

So, I think I have to use kubelogin somehow. What you think?

**Reply by Speeddymon** [reply]
The kubeconfig file I received from an AKS cluster where local accounts were disabled has a command exec field, which specifies arguments for `kubectl` to pass to `kubelogin`. One of these is `--login devicecode` which is responsible for providing a code on the command line that you're supposed to paste into the browser on the computer where `kubectl` is being run. If flux is talking to the kubernetes API on the remote cluster directly without using `kubectl`, then `kubelogin` would never be called even if the binary is in the image, no matter what the kubeconfig contains.

When I remove `kubelogin` from my `$PATH`, and then try to run `kubectl get nodes` or `helm ls -A`, I see this:

```
Unable to connect to the server: getting credentials: exec: executable kubelogin not found

It looks like you are trying to use a client-go credential plugin that is not installed.

To learn more about this feature, consult the documentation available at:
      https://kubernetes.io/docs/reference/access-authn-authz/authentication/#client-go-credential-plugins


kubelogin is not installed which is required to connect to AAD enabled cluster.

To learn more, please go to https://aka.ms/aks/kubelogin
```

For helm the top line above is slightly different, but the remaining lines are identical:

```
Error: Kubernetes cluster unreachable: Get "https://my-server-address:443/version": getting credentials: exec: executable kubelogin not found
```

While, on my flux helm-controller pod, I am getting `"error":"failed to get last release revision: query: failed to query with labels: Unauthorized"`

I'm at a loss for how to connect to these clusters remotely with Flux or how to debug this further

**Reply by makkes** [reply]
Have you tried creating a ServiceAccount? https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#manually-create-an-api-token-for-a-serviceaccount

**Comment by Speeddymon** [comment]
I haven't created one yet (planned to try it today) and it should work fine
as I was able to connect to the remote cluster with the admin kubeconfig
once I reenabled local accounts.

The only issue is that I would either need to manually create the service
account and secret or else have flux or crossplane use the admin
credentials to create the remote service account and secret then have flux
use the secret.

The former is less automated but more secure, and the latter is automated
but still requires the admin local account through the entire remote
cluster's lifecycle for ongoing reconciliation of the service account and
secret.

**Reply by makkes** [reply]
I don't know how you're spinning up the AKS cluster but can you add a step to that process to create the SA and a token Secret? Then you can proceed from there by extracting the token from the Secret and passing it to Flux.

**Comment by Speeddymon** [comment]
Unfortunately I don't think I can. It's a managed AKS cluster deployed with
either terraform or crossplane. If there's a way, I haven't found it yet
but will keep looking.
